File: src/lib/registries/crates.ts
import type { ResolvedPackage } from "../../types.js";

const CRATES_API = "https://crates.io/api/v1";

interface CrateVersion {
  num: string;
  yanked: boolean;
  created_at: string;
}

interface CrateResponse {
  crate: {
    id: string;
    name: string;
    max_version: string;
    repository?: string;
    homepage?: string;
  };
  versions: CrateVersion[];
}

interface CrateVersionResponse {
  version: {
    num: string;
    crate: string;
    yanked: boolean;
  };
}

/**
 * Parse a crates.io package specifier like "serde@1.0.0" into name and version
 */
export function parseCratesSpec(spec: string): {
  name: string;
  version?: string;
} {
  // Handle @ version specifier: serde@1.0.0
  const atIndex = spec.lastIndexOf("@");
  if (atIndex > 0) {
    return {
      name: spec.slice(0, atIndex).trim(),
      version: spec.slice(atIndex + 1).trim(),
    };
  }

  return { name: spec.trim() };
}

/**
 * Fetch crate metadata from crates.io
 */
async function fetchCrateInfo(crateName: string): Promise<CrateResponse> {
  const url = `${CRATES_API}/crates/${crateName}`;

  const response = await fetch(url, {
    headers: {
      Accept: "application/json",
      "User-Agent": "opensrc-cli (https://github.com/opensrc-labs/opensrc)",
    },
  });

  if (!response.ok) {
    if (response.status === 404) {
      throw new Error(`Crate "${crateName}" not found on crates.io`);
    }
    throw new Error(
      `Failed to fetch crate info: ${response.status} ${response.statusText}`,
    );
  }

  return response.json() as Promise<CrateResponse>;
}

/**
 * Fetch specific version info from crates.io
 */
async function fetchCrateVersionInfo(
  crateName: string,
  version: string,
): Promise<CrateVersionResponse> {
  const url = `${CRATES_API}/crates/${crateName}/${version}`;

  const response = await fetch(url, {
    headers: {
      Accept: "application/json",
      "User-Agent": "opensrc-cli (https://github.com/opensrc-labs/opensrc)",
    },
  });

  if (!response.ok) {
    if (response.status === 404) {
      throw new Error(
        `Version "${version}" not found for crate "${crateName}"`,
      );
    }
    throw new Error(
      `Failed to fetch crate version info: ${response.status} ${response.statusText}`,
    );
  }

  return response.json() as Promise<CrateVersionResponse>;
}

/**
 * Extract repository URL from crate metadata
 */
function extractRepoUrl(crate: CrateResponse["crate"]): string | null {
  // Check repository field first
  if (crate.repository && isGitRepoUrl(crate.repository)) {
    return normalizeRepoUrl(crate.repository);
  }

  // Fall back to homepage if it's a git repo
  if (crate.homepage && isGitRepoUrl(crate.homepage)) {
    return normalizeRepoUrl(crate.homepage);
  }

  return null;
}

function isGitRepoUrl(url: string): boolean {
  return (
    url.includes("github.com") ||
    url.includes("gitlab.com") ||
    url.includes("bitbucket.org")
  );
}

function normalizeRepoUrl(url: string): string {
  // Remove trailing slashes and common suffixes
  return url
    .replace(/\/+$/, "")
    .replace(/\.git$/, "")
    .replace(/\/tree\/.*$/, "")
    .replace(/\/blob\/.*$/, "");
}

/**
 * Get available versions sorted by release date (newest first)
 */
function getAvailableVersions(versions: CrateVersion[]): string[] {
  return versions
    .filter((v) => !v.yanked)
    .sort(
      (a, b) =>
        new Date(b.created_at).getTime() - new Date(a.created_at).getTime(),
    )
    .map((v) => v.num);
}

/**
 * Resolve a crate to its repository information
 */
export async function resolveCrate(
  crateName: string,
  version?: string,
): Promise<ResolvedPackage> {
  const info = await fetchCrateInfo(crateName);

  // If version specified, verify it exists
  let resolvedVersion = version || info.crate.max_version;

  if (version) {
    await fetchCrateVersionInfo(crateName, version);
    resolvedVersion = version;
  }

  const repoUrl = extractRepoUrl(info.crate);

  if (!repoUrl) {
    const availableVersions = getAvailableVersions(info.versions)
      .slice(0, 5)
      .join(", ");
    throw new Error(
      `No repository URL found for "${crateName}@${resolvedVersion}". ` +
        `This crate may not have its source published. ` +
        `Recent versions: ${availableVersions}`,
    );
  }

  // Rust crates commonly use v1.2.3 as tags
  const gitTag = `v${resolvedVersion}`;

  return {
    registry: "crates",
    name: crateName,
    version: resolvedVersion,
    repoUrl,
    gitTag,
  };
}

File: packages/opensrc/cli/src/lib/registries/crates.rs
use super::{Registry, ResolvedPackage};
use serde::Deserialize;

const CRATES_API: &str = "https://crates.io/api/v1";

#[derive(Deserialize)]
struct CrateInfo {
    max_version: String,
    repository: Option<String>,
    homepage: Option<String>,
}

#[derive(Deserialize)]
struct CrateResponse {
    #[serde(rename = "crate")]
    krate: CrateInfo,
}

// Only used to validate the version exists
#[derive(Deserialize)]
#[allow(dead_code)]
struct CrateVersionResponse {
    version: serde_json::Value,
}

pub fn parse_crates_spec(spec: &str) -> (String, Option<String>) {
    if let Some(at_idx) = spec.rfind('@') {
        if at_idx > 0 {
            return (
                spec[..at_idx].trim().to_string(),
                Some(spec[at_idx + 1..].trim().to_string()),
            );
        }
    }
    (spec.trim().to_string(), None)
}

fn fetch_crate_info(name: &str) -> Result<CrateResponse, Box<dyn std::error::Error>> {
    let url = format!("{CRATES_API}/crates/{name}");

    let client = reqwest::blocking::Client::new();
    let resp = client
        .get(&url)
        .header("Accept", "application/json")
        .header(
            "User-Agent",
            "opensrc-cli (https://github.com/vercel-labs/opensrc)",
        )
        .send()?;

    if resp.status() == reqwest::StatusCode::NOT_FOUND {
        return Err(format!("Crate \"{name}\" not found on crates.io").into());
    }
    if !resp.status().is_success() {
        return Err(format!("Failed to fetch crate info: {}", resp.status()).into());
    }

    Ok(resp.json()?)
}

fn verify_crate_version(name: &str, version: &str) -> Result<(), Box<dyn std::error::Error>> {
    let url = format!("{CRATES_API}/crates/{name}/{version}");

    let client = reqwest::blocking::Client::new();
    let resp = client
        .get(&url)
        .header("Accept", "application/json")
        .header(
            "User-Agent",
            "opensrc-cli (https://github.com/vercel-labs/opensrc)",
        )
        .send()?;

    if resp.status() == reqwest::StatusCode::NOT_FOUND {
        return Err(format!("Version \"{version}\" not found for crate \"{name}\"").into());
    }
    if !resp.status().is_success() {
        return Err(format!("Failed to fetch crate version info: {}", resp.status()).into());
    }

    // Just verify it parses; we don't need the data
    let _: CrateVersionResponse = resp.json()?;
    Ok(())
}

use super::{is_git_repo_url, normalize_repo_url};

fn extract_repo_url(krate: &CrateInfo) -> Option<String> {
    if let Some(ref repo) = krate.repository {
        if is_git_repo_url(repo) {
            return Some(normalize_repo_url(repo));
        }
    }
    if let Some(ref hp) = krate.homepage {
        if is_git_repo_url(hp) {
            return Some(normalize_repo_url(hp));
        }
    }
    None
}

pub fn resolve_crate(
    name: &str,
    version: Option<&str>,
) -> Result<ResolvedPackage, Box<dyn std::error::Error>> {
    let info = fetch_crate_info(name)?;

    let resolved_version = match version {
        Some(v) => {
            verify_crate_version(name, v)?;
            v.to_string()
        }
        None => info.krate.max_version.clone(),
    };

    let repo_url = extract_repo_url(&info.krate).ok_or_else(|| {
        format!(
            "No repository URL found for \"{name}@{resolved_version}\". \
             This crate may not have its source published."
        )
    })?;

    let git_tag = format!("v{resolved_version}");

    Ok(ResolvedPackage {
        registry: Registry::Crates,
        name: name.to_string(),
        version: resolved_version,
        repo_url,
        repo_directory: None,
        git_tag,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_crates_spec_simple() {
        let (name, version) = parse_crates_spec("serde");
        assert_eq!(name, "serde");
        assert_eq!(version, None);
    }

    #[test]
    fn test_parse_crates_spec_with_version() {
        let (name, version) = parse_crates_spec("serde@1.0.200");
        assert_eq!(name, "serde");
        assert_eq!(version, Some("1.0.200".into()));
    }
}
